cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,389 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "scan.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/scan.h>
+#include <thrust/transform_scan.h>
+
+namespace cudf {
+namespace detail {
+
+template <typename T>
+using pair_type = thrust::pair<T, T>;
+
+/**
+ * @brief functor to be summed over in a prefix sum such that
+ * the recurrence in question is solved. See
+ * G. E. Blelloch. Prefix sums and their applications. Technical Report
+ * CMU-CS-90-190, Nov. 1990. S. 1.4
+ * for details
+ */
+template <typename T>
+class recurrence_functor {
+ public:
+  __device__ pair_type<T> operator()(pair_type<T> ci, pair_type<T> cj)
+  {
+    return {ci.first * cj.first, ci.second * cj.first + cj.second};
+  }
+};
+
+template <typename T>
+class ewma_functor_base {
+ public:
+  T beta;
+  ewma_functor_base(T beta) : beta{beta} {}
+
+  pair_type<T> IDENTITY = {1.0, 0.0};
+};
+
+template <typename T, bool nulls, bool is_numerator>
+class ewma_adjust_functor : public ewma_functor_base<T> {
+  using ewma_functor_base<T>::ewma_functor_base;
+
+ public:
+  using tupletype = std::conditional_t<nulls, thrust::tuple<bool, int, T>, T>;
+
+  __device__ pair_type<T> operator()(tupletype const data)
+  {
+    if constexpr (nulls) {
+      bool const valid = thrust::get<0>(data);
+      int const exp    = thrust::get<1>(data);
+      T const input    = thrust::get<2>(data);
+      T const beta     = this->beta;
+      if constexpr (is_numerator) {
+        if (valid and (exp != 0)) {
+          // The value is non-null, but nulls preceeded it
+          // must adjust the second element of the pair
+
+          return {beta * (pow(beta, exp)), input};
+        } else if (!valid) {
+          // the value is null, carry the previous value forward
+          // "identity operator" is used
+          return this->IDENTITY;
+        } else {
+          return {beta, input};
+        }
+      } else {
+        if (valid and (exp != 0)) {
+          // The value is non-null, but nulls preceeded it
+          // must adjust the second element of the pair
+
+          return {beta * (pow(beta, exp)), 1.0};
+        } else if (!valid) {
+          // the value is null, carry the previous value forward
+          // "identity operator" is used
+          return this->IDENTITY;
+        } else {
+          return {this->beta, 1.0};
+        }
+      }
```

Comment:
How about
```suggestion
      if (!valid) {
        // The value is null, so carry the previous value forward.
        // Return the identity operand for this operator.
        return this->IDENTITY;
      } else {
        T second = [=] () {
          if constexpr(is_numerator) {
            return input;
          } else {
            return 1;
          }
        }()
        if (valid and (exp != 0)) {
          // The value is non-null, but nulls preceded it so
          // we must adjust the second element of the pair
          return {beta * (pow(beta, exp)), second};
        } else {
          // The exp is 0, so we just propagate the value
          return {beta, second};
        }
      }
```

Reply:
At this point even the IIFE just to use `if constexpr` might be overkill. A simple ternary `T const second = is_numerator ? input : 1` ought to be fine, I'd be surprised if the compiler couldn't figure out how to optimize that just as well as the `if constexpr` and there's no concern about the code compiling.

Reply:
Refactored this roughly as such